### DISCUSSION TITLE
Redundant epoch check in dtls_get_bitmap?

### DISCUSSION THREAD

**Discussion by fwh-dc** [openssl/openssl#23555] [General Discussion]
Doesn't the check `rl->unprocessed_rcds.epoch != rl->epoch` always evaluate true?

https://github.com/openssl/openssl/blob/925118e8c3b1041ce7f9840c2d67e7f878123e6b/ssl/record/methods/dtls_meth.c#L92-L100

From the comment it seems more correct to check `pqueue_size(rl->unprocessed_rcds.q) == 0`?

**Comment by mattcaswell** [comment]
I think this check (`rl->unprocessed_rcds.epoch != rl->epoch`) can actually be removed completely.

Things work differently since I refactored the record layer (implemented in 3.2). Prior to that there was one queue of unprocessed records. If you got a record destined for the next epoch you would queue it. When swapping epoch's the first thing you would do is empty the unprocessed records that we'd previously queued and reinject them into the new "current" epoch. In that context this comment makes perfect sense - basically we shouldn't queue new unprocessed records if we haven't finished transferring all the old ones into the current epoch.

After the record layer refactor each epoch has a different record layer object - and each record layer object has its own unprocessed records queue. There will *never* be records in this queue that are "leftover" from the previous epoch and haven't been transferred yet.

**Reply by fwh-dc** [reply]
Ok great, I just wanted to make sure I didn't miss something. I'll add a refactoring proposal then.
